torch/distributions/kl.py

```diff
@@ -0,0 +1,83 @@
+from collections import OrderedDict
+
+from torch.distributions.distribution import Distribution
+
+_KL_REGISTRY = OrderedDict()
+_KL_DISPATCH_TABLE = {}
+
+
+def register_kl(type_p, type_q):
+    """
+    Decorator to register a pairwise function with :meth:`kl_divergence`.
+    Usage::
+
+        @register_kl(Normal, Normal)
+        def kl_normal_normal(p, q):
+            # insert implementation here
+
+    Lookup order is:
+
+    1.  First look for an exact match.
+    2.  Otherwise find the first pair of registered superclasses, in order that
+        functions were registered.
+
+    Args:
+        type_p (type): A subclass of :class:`~torch.distributions.Distribution`.
+        type_q (type): A subclass of :class:`~torch.distributions.Distribution`.
+    """
+    if not isinstance(type_p, type) and issubclass(type_p, Distribution):
+        raise TypeError('Expected type_p to be a Distribution subclass but got {}'.format(type_p))
+    if not isinstance(type_q, type) and issubclass(type_q, Distribution):
+        raise TypeError('Expected type_q to be a Distribution subclass but got {}'.format(type_q))
+
+    def decorator(fun):
+        _KL_REGISTRY[type_p, type_q] = fun
+        _KL_DISPATCH_TABLE.clear()  # reset since lookup order may have changed
+        return fun
+
+    return decorator
+
+
+def _dispatch_kl(type_p, type_q):
+    # Look for an exact match.
+    try:
+        return _KL_REGISTRY[type_p, type_q]
+    except KeyError:
+        pass
+    # Look for the first approximate match.
+    for super_p, super_q in _KL_REGISTRY:
```

Comment:
I guess that's ok as long as we keep the class hierarchy flat (depth <= 2), but we should do something better if we ever plan to have longer inheritance chains

Reply:
Sure, let's see what happens when things get more complex. The current strategy works for our depth=3 hierarchy `Distribution` -> `Gamma` -> `Chi2`. I seems safe since in the worst case a user can explicitly register all desired pairs.

Reply:
I'll try to switch to most-specific match assuming single inheritance...

Reply:
Done.